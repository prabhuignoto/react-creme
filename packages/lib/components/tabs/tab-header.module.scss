@use '../../design/core.scss';
@use '../../design/theme.scss';
@use '../../design/position.scss';
@use '../../design/list.scss';
@use '../../design/animate.scss';

$flat-tabs-height: 45px;
$rounded-tabs-height: 45px;

$sizes: (sm, md, lg);

$dark-mode: theme.$dark-mode;

.tab_headers_wrapper {
  @extend %left;
  position: relative;
  width: 100%;

  &.tab_headers_wrapper-flat::after {
    height: 2px;
  }
}

.tab_headers_wrapper::after {
  @include position.position-abs('bottom center');
  content: '';
  display: block;
  height: 4px;
  width: 100%;

  @if $dark-mode == 'true' {
    background: theme.$charcoal-gray;
  } @else {
    background: theme.$platinum;
  }
}

.tab_headers_flat {
  height: $flat-tabs-height;
}

.tab_headers_rounded {
  height: $rounded-tabs-height;

  &::after {
    bottom: 3px;
    height: 1px;
  }
}

.tab_headers {
  @extend %list-horizontal-left;
  overflow-x: hidden;
  overflow-y: hidden;
  scroll-behavior: smooth;
  width: 90%;
  z-index: 1;
}

.tab_header_control {
  @extend %center;
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
  height: 100%;
  justify-content: space-around;
  margin-left: auto;
  width: 5rem;
}

.tab_header_btn {
  @extend %border-radius;
  @extend %icon-lg;
  color: var(--icons-color, theme.$primary);
  cursor: pointer;

  &:hover {
    background: theme.$secondary;
  }

  &.tab_header_btn_disabled {
    @extend %disabled;
  }
}

.tab_header_btn_left {
  transform: rotate(180deg);
}

.tab_head {
  min-width: 120px;
  position: relative;
  user-select: none;

  @if $dark-mode == 'true' {
    color: theme.$white;
  } @else {
    color: theme.$black;
  }

  &:hover {
    color: theme.$primary;
  }

  &.tab_head_rounded {
    height: 80%;
  }

  &.tab_head_flat {
    height: 80%;
  }

  &.tab_head_disabled {
    @extend %disabled;
  }

  &.tab_head_selected {
    color: theme.$primary;
  }

  &:not(.tab_head_selected) {
    cursor: pointer;
  }

  &:not(:first-child) {
    margin-left: 0.75rem;
  }

  &:not(.tab_head_with_icon) {
    @extend %center;
  }

  &.tab_head_with_icon {
    align-items: center;
    display: grid;
    grid-template-columns: 2.5rem calc(100% - 2rem) 0.5rem;
    justify-items: center;
  }
}

@each $size in $sizes {
  .tab_head_#{$size} {
    .tab_header_label {
      @extend %font-#{$size};
    }

    .tab_head_icon {
      @extend %icon-#{$size};
    }
  }
}

.tab_head_flat {
  &.tab_head_selected::after {
    @include position.position-abs('bottom center');
    background: theme.$primary;
    bottom: -4px;
    content: '';
    display: block;
    height: 3px;
    width: 100%;
  }

  &::after {
    @include position.position-abs('bottom center');
    // background: rgba(theme.$primary-rgb, 70%);
    content: '';
    display: block;
    transition: width 0.25s cubic-bezier(0.075, 0.82, 0.165, 1);
    width: 0;
  }

  &.tab_header_selected::after {
    width: 100%;
  }
}

.tab_header_label {
  @extend %left;
  height: 90%;
  position: relative;
  text-transform: capitalize;
  white-space: nowrap;
  width: 90%;

  &.tab_header_selected {
    color: theme.$tertiary;
    position: relative;
  }

  &:not(.tab_header_label_icon) {
    @extend %center;
    width: 100%;
  }
}

.tab_head_rounded {
  @if $dark-mode == 'true' {
    border: 1px solid theme.$charcoal-gray;
  } @else {
    border: 1px solid theme.$platinum;
  }
  border-bottom: 0;
  border-radius: 4px;
  border-bottom-left-radius: 0;
  border-bottom-right-radius: 0;
  position: relative;
}

.tab_head_rounded.tab_head_selected::after {
  @include position.position-abs('bottom center');
  @if $dark-mode == 'true' {
    background: theme.$black;
  } @else {
    background: theme.$white;
  }
  bottom: -2px;
  content: '';
  display: block;
  height: 3px;
  width: 100%;
  z-index: -1;
}

.tab_head_icon {
  align-items: center;
  display: flex;
  justify-content: center;
  justify-self: center;
  padding-left: 0.2rem;
}
