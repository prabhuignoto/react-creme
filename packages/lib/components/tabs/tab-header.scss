@use '../../design/layout.scss';
@use '../../design/colors.scss';
@use '../../design/position.scss';
@use '../../design/icon.scss';
@use '../../design/list.scss';
@use '../../design/animate.scss';

$flat-tabs-height: 45px;
$rounded-tabs-height: 45px;

.rc-tab-headers-wrapper {
  @extend %left;
  position: relative;
  width: 100%;

  &.rc-tab-headers-wrapper-flat::after {
    height: 3px;
  }

  &::after {
    @include position.position-abs('bottom center');
    background: colors.$platinum;
    content: '';
    display: block;
    width: 100%;
  }
}

.rc-tab-headers-flat {
  height: $flat-tabs-height;

  &::after {
    bottom: 5px;
    height: 3px;
  }
}

.rc-tab-headers-rounded {
  height: $rounded-tabs-height;

  &::after {
    bottom: 3px;
    height: 1px;
  }
}

.rc-tab-headers {
  @extend %list-horizontal-left;
  overflow-x: hidden;
  overflow-y: hidden;
  scroll-behavior: smooth;
  width: 90%;
  z-index: 1;
}

.rc-tab-header-control {
  @extend %center;
  background: colors.$light-bg;
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
  justify-content: space-around;
  margin-left: auto;
  width: 5rem;

  &.rc-tab-header-control-flat {
    height: 70%;
  }

  &.rc-tab-header-control-rounded {
    height: 80%;
  }
}

.rc-tab-header-btn {
  @extend %border-radius;
  @extend %icon-lg;
  color: var(--icons-color, colors.$primary);
  cursor: pointer;

  &:hover {
    background: colors.$secondary;
  }

  &.rc-tab-header-btn-disabled {
    @extend %disabled;
  }
}

.rc-tab-header-btn-left {
  transform: rotate(180deg);
}

.rc-tab-head {
  min-width: 120px;
  position: relative;
  user-select: none;

  &.rc-tab-head-rounded {
    height: 80%;
  }

  &.rc-tab-head-flat {
    height: 70%;
  }

  &.rc-tab-head-disabled {
    @extend %disabled;
  }

  &.rc-tab-head-selected {
    color: colors.$primary;
  }

  &:not(:first-child) {
    margin-left: 0.75rem;
  }

  &:not(.rc-tab-head-with-icon) {
    @extend %center;
  }

  &.rc-tab-head-with-icon {
    display: grid;
    grid-template-columns: 2.5rem calc(100% - 2rem) 0.5rem;
    justify-items: center;
  }
}

.rc-tab-head-flat {
  &.rc-tab-head-selected::after {
    @include position.position-abs('bottom center');
    background: colors.$primary;
    bottom: -2px;
    content: '';
    display: block;
    height: 3px;
    width: 100%;
  }

  &::after {
    @include position.position-abs('bottom center');
    background: rgba(colors.$primary, 70%);
    bottom: -2px;
    content: '';
    display: block;
    height: 3px;
    transition: width 0.25s cubic-bezier(0.075, 0.82, 0.165, 1);
    width: 0;
  }

  &.rc-tab-header-selected::after {
    width: 100%;
  }
}

.rc-tab-header-label {
  @include animate.halo(2px, 2px);
  cursor: pointer;
  height: 90%;
  justify-self: flex-start;
  position: relative;
  text-transform: capitalize;
  white-space: nowrap;
  width: 90%;

  &.rc-tab-header-selected {
    color: colors.$tertiary;
    position: relative;
  }

  &:not(.rc-tab-header-label-icon) {
    @extend %center;
    width: 100%;
  }
}

.rc-tab-head-rounded {
  border: 1px solid colors.$platinum;
  border-bottom: 0;
  border-radius: 4px;
  border-bottom-left-radius: 0;
  border-bottom-right-radius: 0;
  position: relative;
}

.rc-tab-head-rounded.rc-tab-head-selected {
  &::after {
    @include position.position-abs('bottom center');
    background: colors.$white;
    bottom: -2px;
    content: '';
    display: block;
    height: 3px;
    width: 100%;
    z-index: -1;
  }
}

.rc-tab-head-icon {
  align-items: center;
  display: flex;
  height: 1.5em;
  justify-content: center;
  justify-self: center;
  padding-left: 0.2rem;
  width: 1.5em;

  svg {
    height: 100%;
    width: 100%;
  }
}
