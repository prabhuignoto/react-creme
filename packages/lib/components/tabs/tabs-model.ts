import { CSSProperties, ReactNode } from "react";

export interface TabsCommonModel {
  tabStyle?: "flat" | "rounded";
  border?: boolean;
  focusable?: boolean;
}
export interface TabHeadProps extends TabsCommonModel {
  id: string;
  name?: string;
  selected?: boolean;
  disabled?: boolean;
  handleTabSelection: (id: string) => void;
  icon?: ReactNode;
}

export interface TabsModel extends TabsCommonModel {
  children: ReactNode[];
  labels: string[];
  width?: number | string;
  style?: CSSProperties;
  disabledTabs?: string[];
  iconsColor?: string;
  icons?: ReactNode[];
}

export interface TabItemModel {
  name: string;
  id: string;
  selected?: boolean;
  disabled?: boolean;
  content?: ReactNode;
}

export interface TabHeadersModel extends TabsCommonModel {
  items: TabItemModel[];
  handleTabSelection: (id: string) => void;
  icons?: ReactNode[];
}
