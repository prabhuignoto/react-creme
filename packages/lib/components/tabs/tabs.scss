@use "../../design/layout.scss";
@use "../../design/colors.scss";
@use "../../design/list.scss";
@use "../../design/position.scss";
@use "../../design/animate.scss";
@use "../../design/icon.scss";

.rc-tab-panel {
  @extend %center;
  min-height: 100%;
  width: 100%;
}

.rc-tabs {
  @extend %col;
  padding: 0.2rem;
  width: var(--min-width);
}

.rc-tab-panels {
  @extend %list;
  min-height: 200px;
  width: 100%;

  &.rc-panel-border {
    border: 1px solid colors.$platinum;
    border-radius: 4px;
    border-top: 0;
    border-top-left-radius: 0;
    border-top-right-radius: 0;
    margin-left: 0.1rem;
  }
}
  