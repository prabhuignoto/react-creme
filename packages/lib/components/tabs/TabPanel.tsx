import classNames from 'classnames';
import React, { useMemo } from 'react';
import { TabPanelProps } from './tabs-model';
import styles from './tabs.module.scss';

/**
 * TabPanel Component
 *    @property {React.ReactNode} children - The content of the tab panel.
 *    @property {boolean} disabled - Whether the tab panel is disabled.
 *    @property {string} id - The unique identifier for the tab panel.
 * @returns {JSX.Element} The TabPanel component.
 */
const TabPanel: React.FunctionComponent<TabPanelProps> = ({
  children,
  disabled,
  id,
}: TabPanelProps) => {
  // Define the class for the tab panel, adding the 'disabled' class if necessary
  const tabPanelClass = useMemo(() => {
    return classNames(styles.tab_panel, {
      [styles.disabled]: disabled,
    });
  }, [disabled]);

  // Render the tab panel with the appropriate ARIA roles and properties
  return (
    <div
      className={tabPanelClass}
      role="tabpanel"
      id={`rc-tab-panel-${id}`}
      aria-labelledby={`rc-tab-${id}`}
    >
      {children}
    </div>
  );
};

TabPanel.displayName = 'TabPanel';

export { TabPanel };
