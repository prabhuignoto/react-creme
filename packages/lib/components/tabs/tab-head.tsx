import classNames from 'classnames';
import React, { useEffect, useMemo, useRef } from 'react';
import useFocusNew from '../common/effects/useFocusNew';
import { TabHeadProps } from './tabs-model';
import './tabs.scss';

const TabHead: React.FC<TabHeadProps> = React.memo(
  ({
    disabled,
    focusable,
    handleTabSelection,
    id,
    name,
    selected,
    tabStyle,
    icon,
    onFocus,
    parentHasFocus,
  }: TabHeadProps) => {
    const ref = useRef<HTMLDivElement>(null);

    // enable focus outlines
    if (!disabled && focusable) {
      // useFocus(ref, () => handleTabSelection(id));
      useFocusNew(ref);
      // useKey(ref, () => handleTabSelection(id));
    }

    const headerLabelClass = useMemo(() => {
      return classNames('rc-tab-header-label', {
        [`rc-tab-header-${tabStyle}`]: true,
        'rc-tab-header-label-icon': icon,
      });
    }, [icon]);

    const tabHeadClass = useMemo(() => {
      return classNames('rc-tab-head', {
        'rc-tab-head-disabled': disabled,
        'rc-tab-head-selected': selected,
        [`rc-tab-head-${tabStyle}`]: true,
      });
    }, [disabled, selected]);

    const tabHeadIcon = useMemo(() => {
      return classNames('rc-tab-head-icon', {
        'rc-tab-head-icon-selected': selected,
      });
    }, []);

    useEffect(() => {
      if (!selected) {
        return;
      }
      if (selected && parentHasFocus) {
        // ref.current?.focus();
      }
    }, [selected, parentHasFocus]);

    return (
      <div
        key={id}
        className={tabHeadClass}
        onClick={() => !disabled && handleTabSelection(id)}
        role="tab"
        aria-selected={selected}
        aria-controls={`rc-tab-panel-${id}`}
        id={`rc-tab-${id}`}
        onFocus={!parentHasFocus ? onFocus : undefined}
      >
        {icon && <span className={tabHeadIcon}>{icon}</span>}
        <span
          className={headerLabelClass}
          ref={ref}
          tabIndex={!disabled && focusable && selected ? 0 : -1}
        >
          {name}
        </span>
      </div>
    );
  },
  (prev, next) =>
    prev.selected === next.selected &&
    prev.disabled === next.disabled &&
    prev.parentHasFocus === next.parentHasFocus
);

TabHead.displayName = 'TabHead';

export { TabHead };
