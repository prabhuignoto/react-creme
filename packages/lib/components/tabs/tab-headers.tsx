import classNames from 'classnames';
import React, {
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import { ChevronRightIcon } from '../../icons';
import { TabHead } from './tab-head';
import './tab-header.scss';
import { TabHeadersProps } from './tabs-model';
import './tabs.scss';

const TabHeaders: React.FunctionComponent<TabHeadersProps> = ({
  items,
  tabStyle,
  handleTabSelection,
  focusable,
  icons,
}: TabHeadersProps) => {
  const tabHeadersRef = useRef<HTMLUListElement | null>(null);

  const [disableScrollRight, setDisableScrollRight] = useState(false);
  const [disableScrollLeft, setDisableScrollLeft] = useState(false);

  const [headerWidth, setHeaderWidth] = useState<number>(0);

  const [scrollLeftCurrent, setScrollLeftCurrent] = useState<{
    dir: 'left' | 'right';
    value: number;
  }>({ dir: 'left', value: 0 });

  const canShowControls = useMemo(() => {
    if (tabHeadersRef.current) {
      const { clientWidth, scrollWidth } =
        tabHeadersRef?.current as HTMLElement;

      return clientWidth < scrollWidth;
    }
  }, [headerWidth]);

  const scrollLeft = useCallback(() => {
    const tabHeaders = tabHeadersRef.current;
    if (tabHeaders) {
      const { scrollLeft, clientWidth } = tabHeaders;
      const newLeft = Math.max(0, scrollLeft - (clientWidth * 1) / 2);

      tabHeaders.scrollLeft = newLeft;
      setScrollLeftCurrent({ dir: 'left', value: newLeft });
    }
  }, []);

  const scrollRight = useCallback(() => {
    const tabHeaders = tabHeadersRef.current;
    if (tabHeaders) {
      const { scrollLeft, clientWidth } = tabHeaders;
      const newLeft = scrollLeft + (clientWidth * 1) / 2;
      tabHeaders.scrollLeft = newLeft;
      setScrollLeftCurrent({ dir: 'right', value: newLeft });
    }
  }, []);

  const onHeadersRef = useCallback((node) => {
    tabHeadersRef.current = node;
    setHeaderWidth(node?.scrollWidth || 0);
  }, []);

  const tabHeadersWrapperClass = useMemo(() => {
    return classNames('rc-tab-headers-wrapper', {
      [`rc-tab-headers-wrapper-${tabStyle}`]: true,
    });
  }, []);

  const tabHeadersClass = useMemo(() => {
    return classNames('rc-tab-headers', {
      [`rc-tab-headers-${tabStyle}`]: true,
    });
  }, []);

  useEffect(() => {
    if (scrollLeftCurrent.value >= 0 && tabHeadersRef.current) {
      const { dir, value } = scrollLeftCurrent;
      const ref = tabHeadersRef.current;

      if (dir === 'left') {
        setDisableScrollLeft(value === 0);
        setDisableScrollRight(false);
      }

      if (dir === 'right') {
        const { clientWidth, scrollWidth } = ref;
        setDisableScrollRight(value + clientWidth >= scrollWidth);
        setDisableScrollLeft(false);
      }
    }
  }, [scrollLeftCurrent.value]);

  return (
    <header className={tabHeadersWrapperClass}>
      <div className={tabHeadersClass} ref={onHeadersRef} role="tablist">
        {items.map(({ id, name, selected, disabled }, index) => (
          <TabHead
            key={id}
            id={id}
            name={name}
            selected={selected}
            handleTabSelection={handleTabSelection}
            tabStyle={tabStyle}
            disabled={disabled}
            focusable={focusable}
            icon={icons?.[index]}
          />
        ))}
      </div>
      {canShowControls ? (
        <div className="rc-tab-headers-control">
          {!disableScrollLeft && (
            <span
              className="rc-tab-header-btn rc-tab-header-btn-left"
              role="button"
              onClick={scrollLeft}
            >
              <ChevronRightIcon />
            </span>
          )}
          {!disableScrollRight && (
            <span
              className="rc-tab-header-btn"
              role="button"
              onClick={scrollRight}
            >
              <ChevronRightIcon />
            </span>
          )}
        </div>
      ) : null}
    </header>
  );
};

TabHeaders.displayName = 'TabHeaders';

export { TabHeaders };
