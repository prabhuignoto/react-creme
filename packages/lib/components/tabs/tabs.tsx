import classNames from 'classnames';
import { nanoid } from 'nanoid';
import React from 'react';
import {
  CSSProperties,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import { isDark } from '../common/utils';
import { TabHeaders } from './tab-headers';
import { TabPanel } from './TabPanel';
import { TabItemProps, TabsProps } from './tabs-model';
import styles from './tabs.module.scss';

const Tabs: React.FunctionComponent<TabsProps> = ({
  border = false,
  children,
  disabledTabs = [],
  focusable = true,
  icons,
  iconsColor,
  labels,
  style = {},
  tabStyle = 'flat',
  width = '100%',
  activeTab,
  size = 'sm',
  minHeight = 200,
}) => {
  const selectionStart = useRef<number>(-1);
  const [activeTabId, setActiveTabId] = useState<string>('');

  // prepare the data for the tabs
  const items = useRef<TabItemProps[]>(
    Array.isArray(children)
      ? children.map((_, index) => {
          // check if the tab is disabled
          const disabled = disabledTabs.includes(labels[index]);

          // check if the tab can be selected on load
          const selected = activeTab
            ? activeTab === labels[index]
            : index === 0 && !disabled;

          const _id = nanoid();

          if (selected) {
            selectionStart.current = index;
          }

          return {
            content: children[index],
            disabled: disabled,
            id: _id,
            name: labels[index],
            selected,
          };
        })
      : []
  );

  if (!items.current.some(item => item.selected)) {
    const availItems = items.current.filter(item => !item.disabled);

    if (availItems) {
      availItems[0].selected = true;
    }
  }

  // gets the tab content based on the active selection
  const getTabContent = useMemo(() => {
    if (activeTabId) {
      return items.current.filter(
        item => !item.disabled && item.id === activeTabId
      )[0].content;
    } else {
      return null;
    }
  }, [disabledTabs.length, activeTabId]);

  // collection of tab items
  const tabItems = useMemo(() => {
    return items.current
      .filter(tab => !tab.disabled)
      .map(
        ({ id }) =>
          id === activeTabId && (
            <TabPanel key={id} id={id}>
              {getTabContent}
            </TabPanel>
          )
      );
  }, [activeTabId]);

  const visibleTabs = useMemo(() => {
    return items.current.filter(item => labels.indexOf(item.name) > -1);
  }, [labels.length]);

  // handles the tab selection
  const handleTabSelection = useCallback((id: string) => {
    setActiveTabId(id);
  }, []);

  // styles and classes
  const tabsStyle = useMemo(
    () =>
      ({
        ...style,
        '--icons-color': iconsColor,
        '--rc-tabs-min-height': `${minHeight}px`,
        '--rc-tabs-min-width': Number.isInteger(width) ? `${width}px` : width,
      }) as CSSProperties,
    []
  );

  const rcTabsClass = useMemo(
    () =>
      classNames(styles.tabs, {
        [styles.tab_border]: border,
      }),
    []
  );

  const isDarkMode = useMemo(() => isDark(), []);

  const rcPanelsClass = useMemo(() => {
    return classNames(styles.tab_panels, {
      [styles.panel_border]: tabStyle === 'rounded',
      [styles.dark]: isDarkMode,
      [styles[tabStyle]]: true,
    });
  }, []);

  useEffect(() => {
    const selected = items.current.find(item => item.selected);

    if (selected) {
      setActiveTabId(selected.id);
    }
  }, []);

  const handleKeyDown = useCallback(
    (ev: React.KeyboardEvent) => {
      const key = ev.key;
      if (key === 'ArrowLeft' || key === 'ArrowRight') {
        ev.stopPropagation();
        ev.preventDefault();

        const _items = items.current;
        const activeTabIndex = _items.findIndex(item => {
          return item.id === activeTabId;
        });

        if (key === 'ArrowLeft' && activeTabIndex > 0) {
          handleTabSelection(_items[activeTabIndex - 1].id);
        } else if (key === 'ArrowRight' && activeTabIndex < _items.length - 1) {
          handleTabSelection(_items[activeTabIndex + 1].id);
        }
      }
    },
    [activeTabId]
  );

  return (
    <div className={rcTabsClass} style={tabsStyle} onKeyDown={handleKeyDown}>
      <TabHeaders
        items={visibleTabs}
        handleTabSelection={handleTabSelection}
        tabStyle={tabStyle}
        focusable={focusable}
        icons={icons}
        activeTabId={activeTabId}
        size={size}
      />
      <div className={rcPanelsClass}>{tabItems}</div>
    </div>
  );
};

Tabs.displayName = 'Tabs';

export { Tabs };
