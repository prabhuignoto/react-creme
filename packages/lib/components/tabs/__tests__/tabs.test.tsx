import { expect, describe, it } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import { Tabs } from '../tabs';

describe('<Tabs />', () => {
  const labels = ['Tab1', 'Tab2', 'Tab3'];
  const children = [
    <div key="1">Content 1</div>,
    <div key="2">Content 2</div>,
    <div key="3">Content 3</div>,
  ];

  it('renders correctly', () => {
    render(<Tabs labels={labels} children={children} />);

    // Check if all labels are present
    labels.forEach(label => {
      expect(screen.getByText(label)).toBeVisible();
    });

    // Check if the first tab's content is rendered
    expect(screen.getByText('Content 1')).toBeVisible();

    // Check if other tabs' content is not rendered
    expect(screen.queryByText('Content 2')).toBeNull();
    expect(screen.queryByText('Content 3')).toBeNull();
  });

  it('handles tab selection', () => {
    render(<Tabs labels={labels} children={children} />);

    // Select the second tab
    fireEvent.click(screen.getByText('Tab2'));

    // Check if the second tab's content is rendered
    expect(screen.getByText('Content 2')).toBeVisible();

    // Check if other tabs' content is not rendered
    expect(screen.queryByText('Content 1')).toBeNull();
    expect(screen.queryByText('Content 3')).toBeNull();
  });

  it('handles keyboard navigation', () => {
    render(<Tabs labels={labels} children={children} />);

    // Navigate to the second tab with the right arrow key
    fireEvent.keyDown(screen.getByText('Tab1'), { key: 'ArrowRight' });

    // Check if the second tab's content is rendered
    expect(screen.getByText('Content 2')).toBeVisible();

    // Navigate back to the first tab with the left arrow key
    fireEvent.keyDown(screen.getByText('Tab2'), { key: 'ArrowLeft' });

    // Check if the first tab's content is rendered
    expect(screen.getByText('Content 1')).toBeVisible();
  });
});
