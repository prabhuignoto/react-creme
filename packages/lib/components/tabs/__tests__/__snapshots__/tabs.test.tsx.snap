// Vitest Snapshot v1

exports[`Tabs > should render snapshot 1`] = `
<div
  class="_tab_headers_10lhs_42 _flat_10lhs_96"
  role="tablist"
>
  <div
    aria-controls="rc-tab-panel-yWFKJmgxh9ibi-t-CHO52"
    aria-selected="true"
    class="_tab_head_10lhs_1 _tab_head_selected_10lhs_168 _tab_head_flat_10lhs_165 _tab_head_sm_10lhs_5"
    id="rc-tab-yWFKJmgxh9ibi-t-CHO52"
    role="tab"
    style="outline: none; position: relative;"
    tabindex="0"
  >
    <span
      class="_tab_header_label_10lhs_5 undefined"
    >
      one
    </span>
    <span
      class="_focus_ring_4fy2t_1"
      style="width: 6px; height: 6px;"
    />
  </div>
  <div
    aria-controls="rc-tab-panel-F5RMvwqiDgW0KxQAJkvK_"
    aria-selected="false"
    class="_tab_head_10lhs_1 _tab_head_flat_10lhs_165 _tab_head_sm_10lhs_5"
    id="rc-tab-F5RMvwqiDgW0KxQAJkvK_"
    role="tab"
    style="outline: none; position: relative;"
    tabindex="-1"
  >
    <span
      class="_tab_header_label_10lhs_5 undefined"
    >
      two
    </span>
    <span
      class="_focus_ring_4fy2t_1"
      style="width: 6px; height: 6px;"
    />
  </div>
  <div
    aria-controls="rc-tab-panel-P2g513g2mZ0OxLUZJOZXg"
    aria-selected="false"
    class="_tab_head_10lhs_1 _tab_head_flat_10lhs_165 _tab_head_sm_10lhs_5"
    id="rc-tab-P2g513g2mZ0OxLUZJOZXg"
    role="tab"
    style="outline: none; position: relative;"
    tabindex="-1"
  >
    <span
      class="_tab_header_label_10lhs_5 undefined"
    >
      three
    </span>
    <span
      class="_focus_ring_4fy2t_1"
      style="width: 6px; height: 6px;"
    />
  </div>
</div>
`;
